### ISSUE TITLE
Pane history not being saved (GitBash) 

### ISSUE THREAD

**Issue by oshihirii** [microsoft/terminal#11697]
### Windows Terminal version (or Windows build number)

Windows 10 , Version 21H1 (OS Build 19043.1288)
Windows Terminal Preview Version: 1.12.2931.0

### Other Software

Git 2.33.1

### Steps to reproduce

I have three Git Bash panes in my Windows Terminal Preview setup.  

I close Windows Terminal Preview by clicking the `X` button in top right of application.  

When I re-open, my most recent history is not accessible in any of these three panes (ie pressing the up and down arrows on the command line).  

It seems to have saved history from before I installed Windows Terminal Preview.  

I have a `.bash_history` file in `C:\Users\Me` and I can see the history entries in that file.  

I don't seem to have a `.bash_profile` file in that location (as related answers on stack overflow have suggested looking at).

### Actual Behavior

See `Steps to reproduce` section.

**Comment by oshihirii** [comment]
I don't know if it is the correct thing to do, but I created a file called `.bash_profile` at `C:\Users\Me` with this content:

    PROMPT_COMMAND='history -a'  

Per this answer:

https://stackoverflow.com/a/10901227

And now I can access recent command history.

**Comment by zadjii-msft** [comment]
Weird. That sure is interesting. I wonder if the history gets sabed when you `exit` bash, rather than hitting the 'X' button on the tab. I think we've seen that before somewhere actually....

Glad you got this sorted out though!

**Comment by oshihirii** [comment]
just fyi - i tested and if i type `exit` it closes the `pane` (as you suggested - the entries are saved in `.bash_history`).  

however that is not desired behaviour as i'd like to retain all the panes in their defined positions when i next open the app.  

doing the above steps requires resetting up panes when re-opening the app.

**Comment by oshihirii** [comment]
just updating that this is still happening as at 15/04/22.  

i am now running it on a new laptop and again found myself searching for "windows terminal preview not saving history".

**Comment by theHilikus** [comment]
any idea why this workaround is actually needed? having to change your history mode to safe on every command seems overkill. the problem seems to me to be that Windows Terminal is not exiting cleanly the session when you use the UI. if i do `exit` or ctrl+d it is persisted properly but not if i click X

Weird thing is that I have WSL2 configured and that one persists the history, but the one for git bash does not

seems to be similar to https://github.com/microsoft/vscode/issues/39877 where indeed VScode is not exiting cleanly

**Comment by DHowett** [comment]
I'm gonna reopen this for triage again. Thanks!

**Comment by Okeanos** [comment]
@DHowett I had a look at that the problem of Terminal not persisting Git Bash history while ago as well and wrote down my thoughts [here](https://github.com/git-for-windows/git/discussions/4071).

TL;DR: it appears that `git-bash.exe` supplied by Git for Windows (`MinTTY` in particular only) handles signal handling differently than Windows Terminal (or VSCode for that matter) and will properly exit out of `bash.exe`. VSCode additionally suffers from not starting Git Bash as interactive session (which is required for automatic history handling; see microsoft/vscode#151105).

**Comment by oshihirii** [comment]
> I don't know if it is the correct thing to do, but I created a file called `.bash_profile` at `C:\Users\Me` with this content:
> 
> ```
> PROMPT_COMMAND='history -a'  
> ```
> 
> Per this answer:
> 
> https://stackoverflow.com/a/10901227
> 
> And now I can access recent command history.

Just noting, I have switched to another PC and found myself googling this issue again:

> windows terminal preview not saving bash history

The environment is:  
  
 - Windows Terminal Preview:  Version: 1.16.3464.0
 - Windows 11 Home

I got it working by doing the workaround I posted above on 06/11/21. 

Someone down-thumbed that post though, so maybe it's not a good thing to do :cry:

**Comment by oshihirii** [comment]
Anyone got a solution for this?

I just setup Windows Terminal Preview again on new computer, with Git Bash as Default Profile (i use three panes) and there is still no command history saved when I close and re-open WTP (I have enabled WTP setting to restore windows from previous session).  When I try the fix I documented above, Windows Terminal Preview will not open afterwards (just get spinning cursor) - so that fix seems to break things now.
